common/quotas/ratelimiter.go

```diff
@@ -1,145 +0,0 @@
-
-const (
-	_defaultRPSTTL = 60 * time.Second
-	_burstSize     = 1
-)
-
-// TODO: this is likely major overkill, especially now that clock.Ratelimiter exists.
-// Just mutate the limit, it's concurrency-safe and efficient.  And consider a background
-// debouncer only if that proves too costly (stevenl: I suspect it will not be).
-
-// RateLimiter is a wrapper around the golang rate limiter handling dynamic
-// configuration updates of the max dispatch per second. This has comparable
-// performance to the token bucket rate limiter.
-// BenchmarkSimpleRateLimiter-4   	10000000	       114 ns/op (tokenbucket)
-// BenchmarkRateLimiter-4         	10000000	       148 ns/op (this)
-type RateLimiter struct {
-	sync.RWMutex
-	maxDispatchPerSecond *float64
-	goRateLimiter        atomic.Value
-	// TTL is used to determine whether to update the limit. Until TTL, pick
-	// lower(existing TTL, input TTL). After TTL, pick input TTL if different from existing TTL
-	ttlTimer *time.Timer
-	ttl      time.Duration
-	minBurst int
-}
-
-// NewSimpleRateLimiter returns a new rate limiter backed by the golang rate
-// limiter.  This is currently only used in tests.
-func NewSimpleRateLimiter(t testing.TB, rps int) *RateLimiter {
-	t.Helper() // ensure a T has been passed
-	initialRps := float64(rps)
-	return NewRateLimiter(&initialRps, _defaultRPSTTL, _burstSize)
-}
-
-// NewRateLimiter returns a new rate limiter that can handle dynamic
-// configuration updates
-func NewRateLimiter(maxDispatchPerSecond *float64, ttl time.Duration, minBurst int) *RateLimiter {
-	rl := &RateLimiter{
-		maxDispatchPerSecond: maxDispatchPerSecond,
-		ttl:                  ttl,
-		ttlTimer:             time.NewTimer(ttl),
-		minBurst:             minBurst,
-	}
-	rl.storeLimiter(maxDispatchPerSecond)
-	return rl
-}
-
-// UpdateMaxDispatch updates the max dispatch rate of the rate limiter
-func (rl *RateLimiter) UpdateMaxDispatch(maxDispatchPerSecond *float64) {
-	if rl.shouldUpdate(maxDispatchPerSecond) {
-		rl.Lock()
-		rl.maxDispatchPerSecond = maxDispatchPerSecond
-		rl.storeLimiter(maxDispatchPerSecond)
-		rl.Unlock()
-	}
-}
-
-// Wait waits up till deadline for a rate limit token
-func (rl *RateLimiter) Wait(ctx context.Context) error {
-	limiter := rl.goRateLimiter.Load().(clock.Ratelimiter)
-	return limiter.Wait(ctx)
-}
-
-// Reserve reserves a rate limit token
-func (rl *RateLimiter) Reserve() clock.Reservation {
-	limiter := rl.goRateLimiter.Load().(clock.Ratelimiter)
-	return limiter.Reserve()
-}
-
-// Allow immediately returns with true or false indicating if a rate limit
-// token is available or not
-func (rl *RateLimiter) Allow() bool {
-	limiter := rl.goRateLimiter.Load().(clock.Ratelimiter)
-	return limiter.Allow()
-}
-
-// Limit returns the current rate per second limit for this ratelimiter
-func (rl *RateLimiter) Limit() rate.Limit {
-	rl.RLock()
-	defer rl.RUnlock()
-	if rl.maxDispatchPerSecond != nil {
-		return rate.Limit(*rl.maxDispatchPerSecond)
-	}
-	return rate.Inf
```

Comment:
I'm _reasonably_ sure this is not ever used in current code, but to call it out as a caution: you've definitely checked this behavior specifically?

Reply:
Other than Matcher, every use case uses DynamicRateLimiter which has a function returning `float64`, so there's no chance of `nil`. Matcher meanwhile has a default value specified in the config if a poller doesn't provide a rate (as all Decision workers don't provide one) so it won't ever pass `nil`.